nbs/05_preprocessing/py/00-gtex_v8-split_by_tissue.py

```diff
@@ -0,0 +1,322 @@
+# ---
+# jupyter:
+#   jupytext:
+#     cell_metadata_filter: all,-execution,-papermill,-trusted
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.4
+#   kernelspec:
+#     display_name: Python 3 (ipykernel)
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown] tags=[]
+# # Description
+
+# %% [markdown] tags=[]
+# It splits gene expression data from GTEx v8 by tissue and saves a gene id/symbol mapping file.
+
+# %% [markdown] tags=[]
+# # Modules
+
+# %% tags=[]
+import pandas as pd
+from tqdm import tqdm
+
+from clustermatch.utils import simplify_string
+from clustermatch import conf
+
+# %% [markdown] tags=[]
+# # Settings
+
+# %% tags=[]
+OUTPUT_DIR = conf.GTEX["DATA_DIR"] / "data_by_tissue"
+OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
+display(OUTPUT_DIR)
+
+# %% [markdown] tags=[]
+# # Data loading
+
+# %% [markdown] tags=[]
+# ## GTEx v8
+
+# %% [markdown] tags=[]
+# ### Sample metadata
+
+# %% tags=[]
+gtex_sample_attrs = pd.read_csv(
+    conf.GTEX["SAMPLE_ATTRS_FILE"],
+    sep="\t",
+)
+
+# %% tags=[]
+gtex_sample_attrs.shape
+
+# %% tags=[]
+gtex_sample_attrs.head()
+
+# %% [markdown] tags=[]
+# # Get tissues names
+
+# %% tags=[]
+gtex_tissues = gtex_sample_attrs["SMTSD"].unique()
+display(len(gtex_tissues))
+display(gtex_tissues)
+
+# %% [markdown] tags=[]
+# # Get sample IDs for each tissue
+
+# %% tags=[]
+# first, get all sample IDs with expression data
+gtex_all_sample_ids_with_expr_data = set(
+    pd.read_csv(
+        conf.GTEX["DATA_TPM_GCT_FILE"],
+        sep="\t",
+        skiprows=2,
+        nrows=1,
+        usecols=lambda x: x not in ("Name", "Description"),
+    ).columns
+)
+
+# %% tags=[]
+len(gtex_all_sample_ids_with_expr_data)
+
+# %% tags=[]
+list(gtex_all_sample_ids_with_expr_data)[:10]
+
+# %% tags=[]
+# get sample IDs by tissue
+sample_ids_by_tissue = {
+    tissue_name: sorted(
+        list(
+            gtex_all_sample_ids_with_expr_data.intersection(
+                set(
+                    gtex_sample_attrs[gtex_sample_attrs["SMTSD"] == tissue_name][
+                        "SAMPID"
+                    ].tolist()
+                )
+            )
+        )
+    )
+    for tissue_name in gtex_tissues
+}
+
+# %% tags=[]
+assert len(gtex_tissues) == len(sample_ids_by_tissue)
+
+# %% tags=[]
+sample_ids_by_tissue["Whole Blood"][:10]
+
+# %% tags=[]
+# all IDs are unique
+assert all(
+    [
+        len(sample_ids_by_tissue[tissue_name])
+        == len(set(sample_ids_by_tissue[tissue_name]))
+        for tissue_name in sample_ids_by_tissue.keys()
+    ]
+)
+
+# %% [markdown] tags=[]
+# ## Show sample size by tissue
+
+# %% tags=[]
+tissue_sample_size = pd.DataFrame(
+    [{"tissue": k, "sample_size": len(v)} for k, v in sample_ids_by_tissue.items()]
```

Comment:
Yeah, thanks! I love `groupby` and I try to use it whenever I can. I just tried this, but just remembered that `sample_ids_by_tissue` is a dictionary with tissue name as key and a list of sample IDs as values, so first I would need to convert that into a DataFrame to use the `groupby` function. I think my first attempt when I wrote this was to use groupby actually. Since that would be a large dataframe and I don't need to keep the IDs, I prefer here to keep only the final number of IDs per tissue.